gradle-recommended-product-dependencies/src/main/java/com/palantir/gradle/dist/RecommendedProductDependenciesPlugin.java

```diff
@@ -36,18 +45,56 @@ public final void apply(Project project) {
         project.getPluginManager().withPlugin("java", _plugin -> {
             embedResource(project, ext);
             configureManifest(project, ext);
+            validateProductDependencies(project, ext);
         });
     }
 
+    private void validateProductDependencies(Project project, RecommendedProductDependenciesExtension ext) {
+        TaskProvider<ValidateProductDependenciesTask> validateProductDependenciesTask = project.getTasks()
+                .register(
+                        "validateProductDependencies",
+                        ValidateProductDependenciesTask.class,
+                        task -> task.getRecommendedProductDependencies()
+                                .set(ext.getRecommendedProductDependenciesProvider()));
+
+        project.getTasks().named("check").configure(task -> task.dependsOn(validateProductDependenciesTask));
+    }
+
     private void configureManifest(Project project, RecommendedProductDependenciesExtension ext) {
-        TaskProvider<ConfigureProductDependenciesTask> configureProductDependenciesTask = project.getTasks()
-                .register("configureProductDependencies", ConfigureProductDependenciesTask.class, cmt -> {
-                    cmt.setProductDependencies(ext.getRecommendedProductDependenciesProvider());
-                });
+        // Capture the provider during configuration time to avoid serializing non-serializable objects
+        final Provider<?> productDependenciesProvider = ext.getRecommendedProductDependenciesProvider();
 
-        // Ensure that the jar task depends on this wiring task
+        // Directly configure the Jar task's manifest using doFirst
         project.getTasks().withType(Jar.class).named(JavaPlugin.JAR_TASK_NAME).configure(jar -> {
-            jar.dependsOn(configureProductDependenciesTask);
+            jar.doFirst(new Action<Task>() {
```

Comment:
I think we can do a bit better here. 

The semantics are still different from before. But what I believe you want to do is, have your task create a `Manifest` and write it out to file.

Then you configure the jar task:
```java
jarTask.from(yourManifestGeneratingTask.flatMap(YourManifestGeneratingTask::getOutputFile));
```

Reply:
`Manifest#from` merges manifests from other places into the manifest for the Jar task. So you can define what you want, and then merge it in later. It's not as good as having the lazy attributes. But I think it probably works out quite well considering you want to serialise bits here and there and it's probably better as a task output.